### PR TITLE
59 error handling simpler

### LINKED ISSUE #59: [BUG] failing eval_code breaks the shiny app
When eval_code fails quosure is not returned anymore. It means that reactive which fails return error and breaks the whole app. What we actually need is to still return the quosure and keep the error messages (and warnings) per code element to print the proper diagnostics. 
If we decide to drop diagnostics feature some tryCatch is still needed, because reactives still should return quosure and instead of error message we would like to use `validate(q3)` in shiny apps.

```r
q1 <- new_quosure()
q2 <- eval_code(q1, iris2 <- iris)

q3 <- eval_code(q2, stop("elo"))
q3[["iris2"]]
```

```r
r1 <- reactive(new_quosure())
r2 <- reactive(eval_code(r1(), iris2 <- iris))

r3 <- reactive(r2(), stop("elo"))

get_code(r3()) # get_code fails as r3 doesn't return quosure anymore

```

Final acceptance criteria not yet formulated

### PR DESCRIPTION
**Pull request by gogonzo**
closes #59 Please compare with #61 

- when eval_code fails then returns error condition of class (s3) `c("quosure.error", "try-error", "error", "condition")`
- needed to register S3 method for `[[.quosure.error` as S4 couldn't dispatch on the `quosure.error` like other methods (probably it's handled internally which has priority). Class try-error needed to make `validate` stop. Go to `shiny:::isTruthy` to see when validate interrupts the reactive.
- I'm not happy that `req(<quosure.error>)` in q8 doesn't return error like in q13

image

-----
example
```r
ui <- fluidPage(
  fluidRow(
    column(6, h1("Code")),
    column(6, h1("Output"))
  ),
  fluidRow(
    column(6, verbatimTextOutput("q1_code")),
    column(6, verbatimTextOutput("q1_output"))
  ),
  fluidRow(
    column(6, verbatimTextOutput("q2_code")),
    column(6, verbatimTextOutput("q2_output"))
  ),
  fluidRow(
    column(6, verbatimTextOutput("q3_code")),
    column(6, verbatimTextOutput("q3_output"))
  ),
  fluidRow(
    column(6, verbatimTextOutput("q4_code")),
    column(6, verbatimTextOutput("q4_output"))
  ),
  fluidRow(
    column(6, verbatimTextOutput("q5_code")),
    column(6, verbatimTextOutput("q5_output"))
  ),
  fluidRow(
    column(6, verbatimTextOutput("q6_code")),
    column(6, verbatimTextOutput("q6_output"))
  ),
  fluidRow(
    column(6, verbatimTextOutput("q7_code")),
    column(6, verbatimTextOutput("q7_output"))
  ),
  fluidRow(
    column(6, verbatimTextOutput("q8_code")),
    column(6, verbatimTextOutput("q8_output"))
  ),
  fluidRow(
    column(6, verbatimTextOutput("q9_code")),
    column(6, verbatimTextOutput("q9_output"))
  ),
  fluidRow(
    column(6, verbatimTextOutput("q10_code")),
    column(6, verbatimTextOutput("q10_output"))
  ),
  fluidRow(
    column(6, verbatimTextOutput("q11_code")),
    column(6, verbatimTextOutput("q11_output"))
  ),
  fluidRow(
    column(6, verbatimTextOutput("q12_code")),
    column(6, verbatimTextOutput("q12_output"))
  ),
  fluidRow(
    column(6, verbatimTextOutput("q13_code")),
    column(6, verbatimTextOutput("q13_output"))
  )
)

server <- function(input, output, session) {
  q1 <- reactive({
    new_quosure() %>%
    eval_code(code = 'stop("Error message details")')
  })

  output$q1_code <- reactive(
    'q1 <- reactive(new_quosure() %>%
      eval_code(stop("Error message details")))'
  )

  output$q1_output <- renderPrint(q1())

  q2 <- reactive({
    new_quosure() %>%
      eval_code(code = "x <- 0")
  })


  output$q2_code <- reactive(
    'q2 <- reactive(new_quosure() %>%
      eval_code("x <- 0")'
  )

  output$q2_output <- renderPrint(q2())

  q3 <- reactive({
    validate(need(q1(), "q1 is not valid"))
    "hello"
  })

  output$q3_code <- reactive(
    'q3 <- reactive({
    validate(need(q1(), "q1 is not valid"))
    "hello"
    })'
  )

  output$q3_output <- renderPrint(q3())

  q4 <- reactive({
    validate(need(q2(), "q2 is not valid"))
    "hello"
  })


  output$q4_code <- reactive(
    'q4 <- reactive({
     validate(need(q2(), "q2 is not valid"))
    "hello"
    })'
  )

  output$q4_output <- renderPrint(q4())

  q5 <- reactive({
    q1()[["x"]]
  })

  # Note case q2()[["not_valid"]] handled by teal.code issue #57

  output$q5_code <- reactive(
    'q5 <- reactive(q1()[["x"]])'
  )

  output$q5_output <- renderPrint({
    q5()
  })


  q6 <- reactive({
    eval_code(q2(), "w <- y * x")
  })


  output$q6_code <- reactive(
    'q6 <- reactive(eval_code(q2(), "w <- y * x"))'
  )

  output$q6_output <- renderPrint(q6())

  q7 <- reactive({
    eval_code(q1(), "w <- 10")
  })


  output$q7_code <- reactive(
    'q7 <- reactive(eval_code(q1(), "w <- 10")'
  )

  output$q7_output <- renderPrint(q7())

  q8 <- reactive({
    req(q1())
    "hello"
  })


  output$q8_code <- reactive(
    'q8 <- reactive({
      req(q1())
      "hello"
    })'
  )

  output$q8_output <- renderPrint(q8())


  q9 <- reactive({
    join(q2(), q7()) %>% get_var("x")
  })


  output$q9_code <- reactive(
    'q9 <- reactive({
      join(q2(), q7()) %>% get_var("x")
    })'
  )

  output$q9_output <- renderPrint(q9())

  q10 <- reactive({
    join(q2(), q7()) %>% get_code()
  })


  output$q10_code <- reactive(
    'q10 <- reactive({
      join(q2(), q7()) %>% get_code()
    })'
  )

  output$q10_output <- renderPrint(q10())


  q11 <- reactive({
    new_quosure() %>%
      eval_code("x <- 0") %>%
      eval_code("y <- 0") %>%
      eval_code("w <- u * x") %>%
      eval_code("v <- x") %>%
      get_code()
  })


  output$q11_code <- reactive(
    'q11 <- reactive({
      new_quosure() %>%
      eval_code("x <- 0") %>%
      eval_code("y <- 0") %>%
      eval_code("w <- u * x") %>%
      eval_code("v <- x") %>%
      get_code()
    })'
  )

  output$q11_output <- renderPrint(q11())


  q12_a <- reactive({
    validate(need(1 > 4, "Error: 1 is less than 4"))
    new_quosure() %>% eval_code("x <- 1")
  })

  q12_b <- reactive({
    get_code(q12_a())
  })

  output$q12_code <- reactive(
    'q12_a <- reactive({
      validate(need(1 > 4, "Error: 1 is less than 4"))
      new_quosure() %>% eval_code("x <- 1")
    })

    q12_b <- reactive({
      get_code(q12_a())
    })
    '
  )

  output$q12_output <- renderPrint(q12_b())

  q13a <- reactive(stop("error in the reactive"))
  q13b <- reactive({
    req(q13a())
    "hello"
  })
  output$q13_code <- reactive(
    'q13a <- reactive(stop("error in the reactive"))
    q13b <- reactive({
      req(q13a())
      "hello"
    })
    '
  )
  output$q13_output <- renderPrint(q13b())
}

runApp(shinyApp(ui, server))
```

### CODE REVIEW

File: R/get_eval_details.R [hunk 1]
```diff
@@ -101,8 +101,8 @@ get_eval_details_srv <- function(id, chunks) {
             span()
           ),
           `if`(
-            eval_info_summary["errors"] > 0,
-            get_number_in_box(eval_info_summary["errors"], box_color = "#ff0000", char_color = "#ffffff"),
+            eval_info_summary["quosure.error"] > 0,
```
Comment (nikolas-burkoff) [inline R/get_eval_details.R:104]: so does this mean the debug info will work?

File: R/quosure-class.R [hunk 2]
```diff
@@ -25,4 +26,4 @@ setValidity("Quosure", function(object) {
   } else {
     TRUE
   }
-})
+})
```
Comment (nikolas-burkoff) [inline R/quosure-class.R:29]:
```suggestion
})

```

File: R/quosure-errors.R [hunk 3]
```diff
@@ -0,0 +1,2 @@
+# needed to handle try-e
+setOldClass("quosure.error")
```
Comment (nikolas-burkoff) [inline R/quosure-errors.R:2]:
```suggestion
setOldClass("quosure.error")

```

(or move to a new file)

File: R/quosure-get_code.R [hunk 4]
```diff
@@ -12,6 +12,7 @@
 #'
 #' @export
 setGeneric("get_code", function(object) {
+  object
```
Comment (nikolas-burkoff) [inline R/quosure-get_code.R:15]: I'd add a comment here as to why we do this

File: R/quosure-eval_code.R [hunk 5]
```diff
@@ -22,47 +22,55 @@ setGeneric("eval_code", function(object, code, name = "code") {
...
-    object@code <- .keep_code_name_unique(object@code, code)
+setMethod("eval_code", signature = c("Quosure", "character"), function(object, code, name) {
+  checkmate::assert_string(name)
+  if (is.null(names(code))) {
+    code <- paste(code, collapse = "\n")
+    names(code) <- name
+  }
+  id <- sample.int(.Machine$integer.max, size = length(code))
+
+  evaluated_code <- object@code
```
Comment (nikolas-burkoff) [inline R/quosure-eval_code.R:33]: This is what I used to get the trace - not used at the moment

**Review by nikolas-burkoff — COMMENTED** [review]
A few minor comments, I'll look in  detail tomorrow

**Comment by nikolas-burkoff** [comment]
Would you also add class "validation" so that the css for validation gets used in the app compare these:

image

File: R/quosure-get_code.R [hunk 6]
```diff
@@ -20,3 +21,13 @@ setGeneric("get_code", function(object) {
 setMethod("get_code", signature = "Quosure", function(object) {
   object@code
 })
+
+#' @rdname get_code
+#' @export
+setMethod("get_code", signature = "quosure.error", function(object) {
```
Comment (nikolas-burkoff) [inline R/quosure-get_code.R:27]: I wonder if there will be any downstream consequences of this (i.e. if you get_code and then do something with it maybe this should also return a quosure.error but with the extra code in the conditionMessage ?

File: R/get_eval_details.R (hunk 1 above)
Comment (gogonzo) [reply R/get_eval_details.R:104]: my mistake here when changing text by regular expression. I'll fix it

File: R/quosure-get_code.R (hunk 6 above)
Comment (nikolas-burkoff) [reply R/quosure-get_code.R:30]: changed this to return a quosure.error with the trace included

**Comment by gogonzo** [comment]
> and you could have an S3 method for print.quosure.error to print the message out without having the <quosure.error: ...> but I sort of like that as it signposts clearly there's a problem with the quosures

Initially I wanted `quosure` to return red error as it's an error and not the restricted-input-choice. I also like that it returns `<quosure.error ...>`. If I make quosure.error.print I'd probably just turn it red

**Comment by nikolas-burkoff** [comment]
So this is a little more realistic for `teal` at the moment buttons 1 and 2 do nothing when clicked on - I think we should change teal.widgets to output the trace / validation error when you click on them

```
library(teal.code)
library(magrittr)
library(teal.widgets)
library(shiny)

ui <- fluidPage(
  verbatim_popup_ui("rcode_1", "Show R code 1"),
  verbatim_popup_ui("rcode_2", "Show R code 2"),
  verbatim_popup_ui("rcode_3", "Show R code 3")
)

server <- function(input, output, session) {
  # invalid due to code error
  q_1 <- reactive(new_quosure() %>%
    eval_code("x <- 1") %>%
    eval_code("y <- x * w"))


  verbatim_popup_srv(
    id = "rcode_1",
    verbatim_content = reactive(get_code(q_1())),
    title = "rcode_1"
  )

  # invalid due to validate
  q_2 <- reactive({
    validate(need(1 > 4, "1 is less than 4"))
    new_quosure() %>%
      eval_code("x <- 1") %>%
      eval_code("y <- x * 1")
  })

  verbatim_popup_srv(
    id = "rcode_2",
    verbatim_content = reactive(get_code(q_2())),
    title = "rcode_2"
  )

  #valid
  q_3 <- reactive(new_quosure() %>%
    eval_code("x <- 1") %>%
    eval_code("y <- x * 1"))

  verbatim_popup_srv(
    id = "rcode_3",
    verbatim_content = reactive(get_code(q_3())),
    title = "rcode_3"
  )
}

shinyApp(ui, server)

```

File: R/get_eval_details.R (hunk 1 above)
Comment (nikolas-burkoff) [inline R/get_eval_details.R:104]:
```suggestion
            eval_info_summary["errors"] > 0,
```

File: R/get_eval_details.R [hunk 7]
```diff
@@ -101,8 +101,8 @@ get_eval_details_srv <- function(id, chunks) {
...
+            get_number_in_box(eval_info_summary["quosure.error"], box_color = "#ff0000", char_color = "#ffffff"),
```
Comment (nikolas-burkoff) [inline R/get_eval_details.R:105]:
```suggestion
            get_number_in_box(eval_info_summary["errors"], box_color = "#ff0000", char_color = "#ffffff"),
```

**Review by nikolas-burkoff — APPROVED** [review]

**Pull request merged**